---
title: "R_assignment"
author: "Tae-Chun Park"
date: "2024-03-07"
output: html_document
---

### Read data file
```{r}
url1 <- "https://raw.githubusercontent.com/Taechun/R_assignment/master/fang_et_al_genotypes.txt"
fang_et_al_genotypes <- read.table(url1, header = TRUE)
url2 <- "https://raw.githubusercontent.com/Taechun/R_assignment/master/snp_position.txt"
snp_position <- read.table(url2, header = TRUE, fill = TRUE)
```

### Inspect structure and dimention of each file
```{r}
str(fang_et_al_genotypes)
dim(fang_et_al_genotypes)
summary(fang_et_al_genotypes)
str(snp_position)
dim(snp_position)
summary(snp_position)
file.info("fang_et_al_genotypes.txt")$size
file.info("snp_position.txt")$size
```
As we know, the files are so big. 'fang_el_al_genotypes' has 2782 observations, and 986 variables with commends of str and dim and 'snp_position' has 1017 observations, and 15 variations. and the file size of fang_el_al is 11051939 and 82763 for snp_position. And Both files are included characters in each column.

### Join the files

1. As our assignment instructor, I wanted to extract "SNP_ID", "Chromosome" and "Position" from SNP_position file
```{r}
snp_data <- snp_position[,c(1,3,4)]
```

2. After that, I wanted to extract two groups (maize and teosinte) information, respectively, from fang_et_al_genotypes for genotype data.
```{r}
maize <- fang_et_al_genotypes[fang_et_al_genotypes$Group %in% c("ZMMIL", "ZMMLR", "ZMMMR"),]
teosinte <- fang_et_al_genotypes[fang_et_al_genotypes$Group %in% c("ZMPBA", "ZMPIL", "ZMPJA"),]
```

3. Then, I don't need the first three columns of each file.
```{r}
maize_filtered <- maize[,c(4:ncol(maize))]
teosinte_filtered <- teosinte[,c(4:ncol(teosinte))]
```

4. Now, I need to transpose those files
```{r}
trans_maize <- t(maize_filtered)
trans_teosinte <- t(teosinte_filtered)
```

5. Last step, I need to join transposed files and snp_data. Due to the big data, I used head to see a few rows of each data
```{r}
merged_maize <- merge(snp_data, trans_maize)
merged_teosinte <- merge(snp_data, trans_teosinte)
```

### Now I am ready to do the assignment.
- 10 files (1 for each chromosome) with SNPs ordered based on increasing position values and with
missing data encoded by this symbol: ?
```{r}
chr1_maize <- merged_maize[merged_maize$Chromosome == 1, ]
chr1_maize_sorted <- chr1_maize[order(chr1_maize$Position), ]
filtered_maize <- chr1_maize_sorted[1210:1256, ]
filtered_maize_chr1_? <- filtered_maize[apply(filtered_maize[ , 1210:1256], 1, function(x) all (x == "?/?")), ]
write.table(filtered_maize_chr1_?, file = "filtered_maize_chr1", sep = "\t", row.names = FALSE)
```
