---
title: "R_assignment"
author: "Tae-Chun Park"
date: "2024-03-07"
output: html_document
---

### Read data file
```{r}
url1 <- "https://raw.githubusercontent.com/Taechun/R_assignment/master/fang_et_al_genotypes.txt"
fang_et_al_genotypes <- read.table(url1, header = TRUE)
url2 <- "https://raw.githubusercontent.com/Taechun/R_assignment/master/snp_position.txt"
snp_position <- read.table(url2, header = TRUE, fill = TRUE)
```

## Part 1
#### Inspect structure and dimention of each file
```{r}
str(fang_et_al_genotypes)
dim(fang_et_al_genotypes)
str(snp_position)
dim(snp_position)
summary(snp_position)
file.info("fang_et_al_genotypes.txt")$size
file.info("snp_position.txt")$size
```
* With the commands I used, the files are so big. 'fang_el_al_genotypes' has 2782 observations, and 986 variables with commends of str and dim and 'snp_position' has 1017 observations, and 15 variations. and the file size of fang_el_al is 11051939 and 82763 for snp_position. And Both files are included characters in each column.

### Join the files

1. As our assignment instructor, I wanted to extract "SNP_ID", "Chromosome" and "Position" from SNP_position file
```{r}
snp_data <- snp_position[,c(1,3,4)]
```

2. After that, I wanted to extract two groups (maize and teosinte) from fang_et_al_genotypes.
```{r}
maize <- fang_et_al_genotypes[fang_et_al_genotypes$Group %in% c("ZMMIL", "ZMMLR", "ZMMMR"),]
rownames(maize) <- NULL
teosinte <- fang_et_al_genotypes[fang_et_al_genotypes$Group %in% c("ZMPBA", "ZMPIL", "ZMPJA"),]
rownames(teosinte) <- NULL
```

3. Then, I don't need the first three columns of each file.
```{r}
maize_cut <- maize[,c(4:ncol(maize))]
teosinte_cut <- teosinte[,c(4:ncol(teosinte))]
```

4. Now, I need to transpose those files. To do that, I need to change those files to data frame.
```{r}
trans_maize <- t(maize_cut)
trans_teosinte <- t(teosinte_cut)
```

5. Last step, I need to combine those two files
```{r}
trans_maize <- as.data.frame(trans_maize)
trans_maize <- data.frame("SNP_ID"=rownames(trans_maize), trans_maize)
trans_maize$SNP_ID <- row.names(trans_maize)
merged_maize <- merge(snp_data, trans_maize, by = "SNP_ID")
trans_teosinte <- as.data.frame(trans_teosinte)
trans_teosinte <- data.frame("SNP_ID"=rownames(trans_teosinte), trans_teosinte)
trans_teosinte$SNP_ID <- row.names(trans_teosinte)
merged_teosinte <- merge(snp_data, trans_teosinte, by = "SNP_ID")
```

#### Now I am ready to do the assignment.
- 10 files (1 for each chromosome) with SNPs ordered based on increasing position values and with
missing data encoded by this symbol: ?
```{r}
chr_1_data <- merged_maize[merged_maize$Chromosome == 1, ]
chr_1_sorted <- chr_1_data[order(as.numeric(chr_1_data$Position)), ]
missing_rows <- subset(chr_1_sorted, apply(chr_1_sorted[, 4:49], 1, function(x) any(x == "?/?")))
write.table(missing_rows, file = "missing_rows.txt", sep = "\t", quote = FALSE, row.names = TRUE)
```

It takes too long time to repeat chr1 to chr10. So, using a loop to do the operation for chr1 to 10.
```{r}
for (chromosome in 1:10) {
  chr_data <- merged_maize[merged_maize$Chromosome == chromosome, ]
  chr_sorted <- chr_data[order(as.numeric(gsub("[^0-9]", "", chr_data$Position))), ]
  missing_rows <- subset(chr_sorted, apply(chr_sorted[, 4:49], 1, function(x) any(x == "?/?")))
   filename <- paste0("maize_?_rows_chr", chromosome, ".txt")
  write.table(missing_rows, file = filename, sep = "\t", quote = FALSE, row.names = TRUE)
}
```

- 10 files (1 for each chromosome) with SNPs ordered based on decreasing position values and with
missing data encoded by this symbol: -
```{r}
for (chromosome in 1:10) {
  chr_data <- merged_maize[merged_maize$Chromosome == chromosome, ]
  chr_sorted <- chr_data[order(-as.numeric(gsub("[^0-9]", "", chr_data$Position))), ]
  missing_rows <- subset(chr_sorted, apply(chr_sorted[, 4:49], 1, function(x) any(x == "-")))
   filename <- paste0("maize_-_rows_chr", chromosome, ".txt")
  write.table(missing_rows, file = filename, sep = "\t", quote = FALSE, row.names = TRUE)
}
```

- 10 files (1 for each chromosome) with SNPs ordered based on increasing position values and with
missing data encoded by this symbol: ?
```{r}
for (chromosome in 1:10) {
  chr_data <- merged_teosinte[merged_teosinte$Chromosome == chromosome, ]
  chr_sorted <- chr_data[order(as.numeric(gsub("[^0-9]", "", chr_data$Position))), ]
  missing_rows <- subset(chr_sorted, apply(chr_sorted[, 4:49], 1, function(x) any(x == "?/?")))
   filename <- paste0("teosinte_?_rows_chr", chromosome, ".txt")
  write.table(missing_rows, file = filename, sep = "\t", quote = FALSE, row.names = TRUE)
}
```

- 10 files (1 for each chromosome) with SNPs ordered based on decreasing position values and with
missing data encoded by this symbol: -
```{r}
for (chromosome in 1:10) {
  chr_data <- merged_teosinte[merged_teosinte$Chromosome == chromosome, ]
  chr_sorted <- chr_data[order(-as.numeric(gsub("[^0-9]", "", chr_data$Position))), ]
  missing_rows <- subset(chr_sorted, apply(chr_sorted[, 4:49], 1, function(x) any(x == "-")))
   filename <- paste0("teosinte_-_rows_chr", chromosome, ".txt")
  write.table(missing_rows, file = filename, sep = "\t", quote = FALSE, row.names = TRUE)
}
```

## Part 2
#### SNP per chromosome - maize
```{r}
library(ggplot2)
maize_counts <- table(merged_maize$Chromosome)
teosinte_counts <- table(merged_teosinte$Chromosome)
df <- data.frame(
  Chromosome = as.numeric(names(maize_counts)),
  Maize = as.numeric(maize_counts),
  Teosinte = as.numeric(teosinte_counts)
)
df_long <- tidyr::pivot_longer(df, cols = c(Maize, Teosinte), 
                               names_to = "Species", values_to = "Count")
ggplot(df_long, aes(x = factor(Chromosome), y = Count, fill = Species)) +
  geom_bar(stat = "identity", position = "dodge") +
  labs(x = "Chromosome", y = "Number of SNPs", fill = "Species") +
  theme_minimal() +
  theme(legend.position = "top")
```
The distribution of SNPs showed the highest tendency on chr1 and the lowest on chr10. However, chr5 exhibited a distribution similar to chr2, and chr7 showed a distribution similar to chr4. Additionally, the distribution of SNPs was similar across each chromosome in the two species.

#### Missing data and amount of heterozygosity

